lib/src/import_path_base.dart

```diff
@@ -0,0 +1,283 @@
+    return rootUri.endsWith('/') ? rootUri : '$rootUri/';
+  }
+
+  /// Return the search root to [strip] from the displayed import paths.
+  /// If [strip] is `false` returns `null`.
+  /// See [searchRoot].
+  String? get stripSearchRoot => strip ? searchRoot : null;
+
+  /// Prints a message/text.
+  /// - Called by [execute].
+  /// - See [messagePrinter].
+  void printMessage(Object? m) => messagePrinter(m);
+
+  /// Executes the import search and prints the results.
+  /// - If `dots` is `true` it prints the tree in `dots` style
+  /// - See [printMessage] and [ASCIIArtTree].
+  Future<ASCIIArtTree?> execute({bool dots = false}) async {
+    if (!quiet) {
+      printMessage('» Search entry point: $from');
+
+      if (strip) {
+        printMessage(
+            '» Stripping search root from displayed imports: $searchRoot');
+      }
+
+      if (findAll) {
+        printMessage(
+            '» Searching for all import paths for `$importToFind`...\n');
+      } else {
+        printMessage(
+            '» Searching for the shortest import path for `$importToFind`...\n');
+      }
+    }
+
+    var tree = await search(dots: dots);
+
+    if (tree != null) {
+      var treeText = tree.generate();
+      printMessage(treeText);
+    }
+
+    if (!quiet) {
+      if (tree == null) {
+        printMessage(
+            '» Unable to find an import path from $from to $importToFind');
+      } else {
+        var totalFoundPaths = tree.totalLeafs;
+        if (totalFoundPaths > 1) {
+          printMessage(
+              '» Found $totalFoundPaths import paths from $from to $importToFind\n');
+        }
+      }
+    }
+
+    return tree;
+  }
+
+  /// Performs the imports search and returns the tree.
+  /// - If [dots] is `true` uses the `dots` style for the tree.
+  /// - See [ASCIIArtTree].
+  Future<ASCIIArtTree?> search({bool dots = false}) async {
+    var foundPaths = await searchPaths();
+    if (foundPaths == null || foundPaths.isEmpty) return null;
+
+    var asciiArtTree = ASCIIArtTree.fromPaths(
+      foundPaths,
+      stripPrefix: stripSearchRoot,
+      style: dots ? ASCIIArtTreeStyle.dots : ASCIIArtTreeStyle.elegant,
+    );
+
+    return asciiArtTree;
+  }
+
+  late PackageConfig _packageConfig;
+  late String _generatedDir;
+
+  /// Performs the imports search and returns the found import paths.
+  Future<List<List<String>>?> searchPaths() async {
+    var currentDir = Directory.current;
+    _packageConfig = (await findPackageConfig(currentDir))!;
+    _generatedDir = p.join('.dart_tool/build/generated');
+
+    var root = from.scheme == 'package' ? _packageConfig.resolve(from)! : from;
+
+    var foundPaths = _searchImportPaths(root, stripSearchRoot: stripSearchRoot);
+    if (foundPaths.isEmpty) {
+      return null;
+    }
+
+    if (!findAll) {
+      foundPaths.sort((a, b) => a.length.compareTo(b.length));
+      var shortest = foundPaths.first;
+      return [shortest];
```

Comment:
The idea was to use the same code for shortest path and all paths. But the optimized code is small, and we can also have the optimized version too.

Reply:
We could have shared code still (but this algorithm wouldn't work). In particular an iterative algorithm makes it much easier to bail out early (and we would need to do a BFS instead of a DFS so we know we hit the shortest path first and then its just a flag to immediately bail on the first solution found).

Reply:
I will reimplement the search algorithm to be BFS and not recursive, producing all the matching paths of the import graph...